File: leetcode/go/SearchInsertPosition.go
package main

import (
	"fmt"
)

type Test struct {
	nums   []int
	target int
	exp    int
}

var tests = []Test{
	{[]int{1, 3, 5, 6}, 5, 2},
	{[]int{1, 3, 5, 6}, 2, 1},
	{[]int{1, 3, 5, 6}, 7, 4},
}

func searchInsert(nums []int, target int) int {
	i, j := 0, len(nums)
	for i < j {
		mid := (i + j) / 2
		if nums[mid] == target {
			return mid
		} else if nums[mid] < target {
			i = mid + 1
		} else {
			j = mid
		}
	}
	return i
}

func main() {
	fmt.Println("Running...")
	for _, t := range tests {
		res := searchInsert(t.nums, t.target)
		if res == t.exp {
			fmt.Println("Success")
		} else {
			fmt.Printf("Failure: %v != %v\n", res, t.exp)
		}
	}
}

File: leetcode/rust/SearchInsertPosition.rs
struct Test(&'static [i32], i32, i32);

const TESTS: [Test; 3] = [
  Test(&[1,3,5,6], 5, 2),
  Test(&[1,3,5,6], 2, 1),
  Test(&[1,3,5,6], 7, 4),
];

fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
  let mut i: i32 = 0;
  let mut j: i32 = nums.len() as i32;
  while i < j {
    let mid: i32 = (i + j) / 2;
    if nums[mid as usize] == target {
      return mid;
    } else if nums[mid as usize] < target {
      i = mid + 1;
    } else {
      j = mid;
    }
  }
  return i;
}

fn main() {
  println!("Running...");
  for Test(input, target, output) in TESTS {
    let v = input.to_vec();
    let res = search_insert(v, target);
    if res == output {
      println!("Success");
    } else {
      println!("Failure: {} != {}", res, output);
    }
  }
}
